src/main/java/org/truffleruby/language/arguments/ReadArgumentsNode.java

```diff
@@ -0,0 +1,523 @@
+        this.expected = expected;
+        this.needsExpandedHash = needsExpandedHash;
+        this.staticArity = staticArity;
+        this.keywordParameterFrameSlots = keywordParameterFrameSlots;
+        this.readUserKeywordsHashNode = new ReadUserKeywordsHashNode(staticArity.getRequired());
+        this.checkKeywordArgumentNodes = checkKeywordArgumentNodes;
+        this.localAssignmentReadNode = localAssignmentReadNode;
+        this.writePreArgumentNode = writePreArgumentNode;
+        this.readPreArgumentNodes = readPreArgumentNodes;
+        this.writeLocalAssignmentNodes = writeLocalAssignmentNodes;
+        this.preOptNodes = preOptNodes;
+
+        if (EMPTY_ARGUMENTS_HASH == null) {
+            EMPTY_ARGUMENTS_HASH = RubyLanguage.getCurrentLanguage().symbolTable.getSymbol("empty arguments hash");
+        }
+    }
+
+    public abstract Object execute(VirtualFrame frame, KeywordDescriptor descriptor);
+
+    @Specialization
+    protected Object empty(VirtualFrame frame, EmptyKeywordDescriptor descriptor) {
+        clearFrameSlots(frame);
+        checkArity(frame, EmptyKeywordDescriptor.EMPTY);
+        unloadSelf(frame);
+        unloadBlockKWArgs(frame);
+        unloadRequiredPositionals(frame, EmptyKeywordDescriptor.EMPTY);
+        unloadLocalAssignmentReadNode(frame, EmptyKeywordDescriptor.EMPTY);
+        unloadRestArgs(frame, EmptyKeywordDescriptor.EMPTY);
+        unloadPostArgument(frame, EmptyKeywordDescriptor.EMPTY);
+        unloadOptionalArgs(frame);
+        unloadKeyRestArgs(frame, EmptyKeywordDescriptor.EMPTY);
+        unloadDefaultArgs(frame, EmptyKeywordDescriptor.EMPTY);
+        return null;
+    }
+
+    @ExplodeLoop
+    @Specialization(guards = "descriptor == cachedDescriptor", limit = "4")
+    protected Object cached(VirtualFrame frame, NonEmptyKeywordDescriptor descriptor,
+            @Cached("descriptor") NonEmptyKeywordDescriptor cachedDescriptor,
+            @Cached(value = "getSlots(cachedDescriptor)") WriteFrameSlotNode[] descriptorSlots) {
+        clearFrameSlots(frame); // TODO only those not in the descriptor
+        expandKeywordHash(frame, cachedDescriptor);
+        checkArity(frame, cachedDescriptor);
+        unloadSelf(frame);
+        unloadBlockKWArgs(frame);
+        unloadRequiredPositionals(frame, cachedDescriptor);
+        unloadLocalAssignmentReadNode(frame, cachedDescriptor);
+        unloadRestArgs(frame, cachedDescriptor);
+        unloadPostArgument(frame, cachedDescriptor);
+        unloadOptionalArgs(frame);
+        unloadKeyRestArgs(frame, cachedDescriptor);
+
+        // For each keyword in the descriptor, store its value in a local, as long as it's expected.
+
+        for (int n = 0; n < cachedDescriptor.getLength(); n++) {
+            final WriteFrameSlotNode frameSlot = descriptorSlots[n];
+            if (frameSlot != DUMMY) {
+                Object lastArgument = RubyArguments.getKeywordArgumentsValue(frame, n, cachedDescriptor);
```

Comment:
Is this safe if e.g. the caller passed us a RubyHash? Won't we read too far in the frame args?